cc/mcts_node.h

```diff
@@ -143,6 +146,20 @@ class MctsNode {
 
   // Number of virtual losses on this node.
   int num_virtual_losses_applied = 0;
+
+  // Each position contains a Zobrist hash of its stones, which can be used for
+  // superko detection. In order to accelerate superko detection, caches of all
+  // ancestor positions are added at regular depths in the search tree. This
+  // reduces superko detection time complexity from O(N) to O(1).
+  //
+  // If non-null, superko_cache contains the Zobrist hash of all positions
+  // played to this position, including position.stone_hash().
+  // If null, clients should determine whether a position has appeared before
+  // during the game by walking up the tree (via the parent pointer), checking
+  // the position.stone_hash() of each node visited, until a node is found that
+  // contains a non-null superko_cache.
+  using SuperkoCache = absl::flat_hash_set<zobrist::Hash>;
+  std::unique_ptr<SuperkoCache> superko_cache;
```

Comment:
This is going to use a fair bit more memory.

At depth = 200
1/8 * 200 * 8bytes = 200 bytes more per node

Reply:
The important factor is the depth of the tree search, since the tree above the current position is essentially a linked list.

With a branching factory of 1.5, a search depth of 200 would be ~10^35 reads, at which point we probably have bigger problems than a 10% increase in memory consumption.

Reply:
It's not search depth that matters but number of moves played in the game

Reply:
How so?

Every time we play a move, we delete all the child nodes in the search tree except those in the subtree of the move that was played.

Ignoring the depth of the search tree, a game of 200 moves will consequently only hold 200/8 == 50 caches in the ancestor nodes of the final position's node. That's 50*100*8 == 40kB of data, since the average size of the cache will be 100.

Am I missing something? I'd like to know if my estimations are way off base!

Reply:
Q: If we're in move 200 of the game, and there are 25 caches spaced out at moves 8, 16, 24... then the caches from 8...192 are unused because all tree searches from move 201+ will back up to 200 and then not look any further. So it seems like we shouldn't worry about those caches.

Instead, I'm wondering if we branch by factor 1.5, then at 10,000 reads, that would imply a search depth of ~20, with the number of nodes at search depth 8 being ~25 nodes, and the number of nodes at search depth 16 being ~650. Do each of the search nodes at depth 16 get their own cache?